jurity/recommenders/interlist_diversity.py

```diff
@@ -0,0 +1,214 @@
+    """
+
+    # Map each user_id to (0, n_users)
+    users = list(df[user_id_column].unique())
+    user_map = dict(zip(users, range(len(users))))
+
+    # Map each user_id to (0, n_items)
+    items = list(df[item_id_column].unique())
+    item_map = dict(zip(items, range(len(items))))
+
+    # Update user_id, item_id values
+    user_ids = df[user_id_column].map(user_map).values
+    item_ids = df[item_id_column].map(item_map).values
+
+    # Convert to csr matrix
+    csr_matrix = sp.coo_matrix((np.ones(len(user_ids)), (user_ids, item_ids))).tocsr()
+
+    return csr_matrix
+
+
+def interlist_diversity(predicted_results: pd.DataFrame, click_column: str, k: int,
+                        user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id,
+                        sample_size: float = None, seed: int = Constants.default_seed,
+                        chunk_size: int = 1000) -> Tuple[float, int]:
+    """
+        Calculate inter-list diversity metric:
+
+            Inter-list-diversity = 1 - average(cosine_similarity(R_{u_i}, R_{u_j})),
+            where R_{u_i} is the binary indicator vector representing provided recommendations for user u_i, i \ne j.
+
+        Parameters
+        ----------
+        predicted_results: pd.DataFrame
+            Recommendations data frame with (user_id, item_id, score) in each row.
+        k: int
+            Top-k recommendations to consider.
+        user_id_column: str
+            User id column name.
+        item_id_column: str
+            Item id column name.
+        click_column: str
+            Recommendation score column name.
+        sample_size: float
+            Proportion of users to randomly sample for evaluation.
+        seed : int
+            The seed used to create random state.
+        chunk_size: int
+            Chunk size to limit memory usage.
+
+        Returns
+        -------
+        Inter-list diversity metric, number of unique users as the support to get the metric
+        """
+
+    # Sample users
+    if sample_size is not None:
+        df = sample_users(predicted_results, user_id_column, sample_size, seed)
+    else:
+        df = predicted_results
+
+    # Sort by user and score
+    df = df.sort_values(click_column, ascending=False).groupby(user_id_column).head(k)
+
+    # Given user/item id maps, create sparse matrix.
+    sparse_matrix = tocsr(df, user_id_column, item_id_column)
+
+    # Get pairwise cosine similarities
+    similarities_sum = 0
+    if chunk_size == 1:
+        num_chunks = sparse_matrix.shape[0]
+    else:
+        num_chunks = sparse_matrix.shape[0] // chunk_size + 1
+    for i in range(num_chunks):
+        similarities = cosine_similarity(sparse_matrix[i * chunk_size:(i + 1) * chunk_size], sparse_matrix,
+                                         dense_output=False)
+        similarities_sum += similarities.sum()
+    similarities_sum = (similarities_sum - sparse_matrix.shape[0]) / 2.0
```

Comment:
Can you explain why are you substracting the size of matrix and dividing by two? I don't immediately follow as you use the sum of `cosine_similarity` function outputs for each chunk and then do the average below.

Reply:
Once the pairwise cosine similarity matrix is calculated, on the diagonal, pairs `i, i` would have similarity to be 1, so I would like to subtract these 1s. Then I only want to reserve the unique pairs `i, j` and `i<j`, so the sum gets divided by two. Calculating similarities with all pairs requires too much memory, so we can calculate in chunks. So I am calculating similarities [row_1] to [row_1, row_2], and then [row_2] to [row_1, row_2], instead of getting them all together.